asn1scala/src/main/scala/asn1scala/asn1jvm_Bitstream.scala

```diff
@@ -1656,9 +1657,25 @@ case class BitStream private [asn1scala](
       // TODO: The C code does indeed not reset the stream to the original position
       // in case we return 0 or 1, but doesn't it make sense to do so?
       // if !ret then
-      currentByte = tmpByte
-      currentBit = tmpBit
 
+      ghostExpr(check(BitStream.invariant(currentBit, currentByte, buf.length)))
+
+      // SAM: Here the issue is that the 2 assignments are not atomic, so the invariant can be temporarily violated
+      // For this reason it'd be better to have it as pre- and post condition everywhere rather than relying on the `require`
+      // Otherwise, we could store currentByte and currentBit in a tuple 
+      if(tmpByte == this.buf.length) {
+         ghostExpr({
+            check(this.currentBit == 0)
+            check(tmpBit == 0)
+            check(BitStream.invariant(tmpBit, currentByte, buf.length))
+         })
+         currentBit = tmpBit
+         currentByte = tmpByte
+      } else {
+         currentByte = tmpByte
+         currentBit = tmpBit
+      }
```

Comment:
@mario-bucev this is a bit annoying ^^
We should have the invariant as pre- and post condition

Reply:
IIRC this function is used in one function that is broken even in the original C code, so for now I think we can just ignore it

Reply:
Too bad I spent time verifying it :( haha

Reply:
my bad, sorry for that!